### ISSUE TITLE
open edX: logout doesn't work

### ISSUE THREAD

**Issue by pdpinch** [mitodl/mitxonline#57]
### Steps to Reproduce

1. Login and visit any page on lms-qa.mitxonline.mit.edu
2. Choose "Sign Out" from the user menu in the top right

### Expected Behavior

- User is logged out and then returned to the "home" page at https://mitxonline-rc.mitxonline.mit.edu/

### Actual Behavior

- User is directed to https://lms-qa.mitxonline.mit.edu/logout where there is message that says the user is logged out. However, after a moment, the user is directed to https://lms-qa.mitxonline.mit.edu/dashboard and is logged back in again. 

I'm not sure how we addressed this in xPRO. It seems like it directs the user to xpro.mit.edu/logout after having logged them out of courses.xpro.mit.edu

**Comment by pdpinch** [comment]
It looks like there were two passes at fixing this in mitxpro: 

1. https://github.com/mitodl/mitxpro/pull/139
2. https://github.com/mitodl/mitxpro/issues/1931

**Comment by arslanashraf7** [comment]
Update:

While working on the issue I noticed that the master branch of our edx-platform fork doesn't have some commits that we added for xPRO in the koa branch [xpro/koa](https://github.com/mitodl/edx-platform/tree/xpro/koa) to fix some redirection issues. 
So, I've created a new branch [mitx-online/lilac](https://github.com/mitodl/edx-platform/tree/mitx-online/lilac), that's based on the current master and it has that redirection fixing commits from [xpro/koa](https://github.com/mitodl/edx-platform/tree/xpro/koa).

Additionally, These are the environment settings/vars that we might consider looking at:

**In MITxOnline:**
- `LOGOUT_REDIRECT_URL` is set to something like `https://lms-qa.mitxonline.mit.edu/logout`

**In LMS:**
-  `IDA_LOGOUT_URI_LIST` should have an entry e.g something like `https://mitxonline-rc.mitxonline.mit.edu/logout`
- `LOGOUT_REDIRECT_URL` should have an entry e.g. something like `https://mitxonline-rc.mitxonline.mit.edu`
- `MARKETING_SITE_BASE_URL` this should be something like `https://mitxonline-rc.mitxonline.mit.edu`

However, I wasn't able to test it locally due to some webpack related issue that broke my devstack and I'll be testing it as soon as my devstack is up and will be posting more updates.

Meanwhile, If we want we can give it a try deploying [mitx-online/lilac](https://github.com/mitodl/edx-platform/tree/mitx-online/lilac) on QA and checking the above mentioned env variables have appropriate values.


CC: @pdpinch @mitodl/devops

**Comment by blarghmatey** [comment]
So, a problem with this approach is that we are explicitly tracking the master branch of edx-platform for this installation rather than deploying a named release. This means that adding custom patches becomes a more complicated endeavor, so if we need custom code in edx-platform for the MITx Online installation we'll need to either:
- get the changes merged upstream and consume them that way
- add the behavior to an existing edx plugin or create a new one that we can install at build time

The reason for us to be tracking master is that it allows us to stay current with what edx.org is offering in terms of features, as well as removing the need for taking down time every 6 months for a major upgrade.

**Comment by arslanashraf7** [comment]
@blarghmatey Thanks for adding the details on keeping up with the master. 

So, adding to what you said earlier and about the [mitx-online/lilac](https://github.com/mitodl/edx-platform/tree/mitx-online/lilac) branch that I created to introduce some changes, It turns out that part of the changes from the branch has already been merged in the upstream https://github.com/edx/edx-platform/pull/26760 and to utilize these we need to:
- Update the master of our fork to at least include [this commit](https://github.com/edx/edx-platform/commit/4a2549f49af44c12f03d957190948800f8614450) but it's going to be a part of [another ticket](https://github.com/mitodl/mitxonline/issues/102) and not this one.
- Decide if the case [here](https://github.com/mitodl/mitxpro/pull/2103) is general enough to take it upstream or if we could manage it inside our theme?

But as for the current logout issue that I was debugging, I saw a couple of PRs(https://github.com/edx/edx-platform/pull/23194 & https://github.com/edx/edx-platform/pull/23229) that introduced some changes in the logout view, and now I believe `self.oauth_client_ids` list is never initialized and no `logout` URL will be picked from `redirect_uris` from the created DOT(Django OAuth Toolkit) applications. So, what we could do here is:

- Add something like `https://mitxonline-rc.mitxonline.mit.edu/logout/` in `IDA_LOGOUT_URI_LIST `, something similar is being done for xPRO [here](https://github.com/mitodl/salt-ops/blob/ab5b7491de4fd5b63a432123b2a0d9f6b4919bb5/pillar/edx/ansible_vars/xpro.sls#L32). This would make sure that edX hits `/logout` of mitxonline and the issue will be fixed. [Ref for DOT apps in edx-platform](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/user_authn/views/logout.py#L136-L137).
- Continue investigating more on the upstream PRs (https://github.com/edx/edx-platform/pull/23194 & https://github.com/edx/edx-platform/pull/23229) and eventually look for possible fixes and put upstream PRs. (If these changes were not intentional)


CC: @pdpinch

**Comment by arslanashraf7** [comment]
Update:

After updating the `IDA_LOGOUT_URI_LIST` with logout URL based on the marketing site fixed it.

[The Reference](https://github.com/mitodl/ol-infrastructure/commit/52324feb6b83137cab4d15a1aa30ba72e865a3ee) can be seen here.

**Comment by blarghmatey** [comment]
@briangrossman can you confirm that this is working as expected on RC and production?

**Comment by briangrossman** [comment]
For the most part, looks like it's working as expected. There's a little oddity when you try to access an internal edX page directly when not logged in (noted below), but otherwise looks good. 

Testing results:

RC
----

**Test 1 - Success**
- While logged in, went to LMS page: https://mfe-qa.mitxonline.mit.edu/learn/course/course-v1:ShiraU+14.750x_SF+3T2021/home
- Clicked `Sign Out`
- Redirected to LMS log out page, then MITx Online login page

**Test 2 - Success**
- From the dashboard: https://courses-qa.mitxonline.mit.edu/dashboard, clicked `Sign Out`
- Redirected to LMS log out page, then MITx Online login page

**Issue**
Not sure if this is related, but...
- If I'm not logged in and I try to access [the LMS dashboard](https://courses-qa.mitxonline.mit.edu/dashboard), I am redirected to the MITx Online sign in page. This seems like correct behavior
- If I'm not logged in and I try to access an internal course page [example](https://mfe-qa.mitxonline.mit.edu/learn/course/course-v1:ShiraU+14.750x_SF+3T2021/home), I get a blank page and errors (below). This page should also redirect to the MITx Online login:
```
home:1 Access to XMLHttpRequest at 'https://rc.mitxonline.mit.edu/oauth2/authorize/?client_id=FV0V9m32EpmIwGaU3MeEruCHWACHgSvolKwpApfi&redirect_uri=https%3A%2F%2Fcourses-qa.mitxonline.mit.edu%2Fauth%2Fcomplete%2Fmitxpro-oauth2%2F%3Fredirect_state%3DfxksXc8prcOY7207AeX5xrPr1UE80b5C&state=fxksXc8prcOY7207AeX5xrPr1UE80b5C&response_type=code&scope=user%3Aread' (redirected from 'https://courses-qa.mitxonline.mit.edu/login_refresh') from origin 'https://mfe-qa.mitxonline.mit.edu' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
rc.mitxonline.mit.edu/oauth2/authorize/?client_id=FV0V9m32EpmIwGaU3MeEruCHWACHgSvolKwpApfi&redirect_uri=https%3A%2F%2Fcourses-qa.mitxonline.mit.edu%2Fauth%2Fcomplete%2Fmitxpro-oauth2%2F%3Fredirect_state%3DfxksXc8prcOY7207AeX5xrPr1UE80b5C&state=fxksXc8prcOY7207AeX5xrPr1UE80b5C&response_type=code&scope=user%3Aread:1 Failed to load resource: net::ERR_FAILED
undefined_license_key:1 Failed to load resource: the server responded with a status of 403 (Forbidden)
```

Prod
----

**Test 1 - Success**
- Logged in to MITx Online
- Accessed the LMS dashboard: https://courses.mitxonline.mit.edu/dashboard
- Clicked `Sign Out`
- Redirected to LMS log out page, then MITx Online login page

**Comment by arslanashraf7** [comment]
@briangrossman as per your comment:

>Issue
Not sure if this is related, but...
If I'm not logged in and I try to access the LMS dashboard, I am redirected to the MITx Online sign in page. This seems like correct behavior
If I'm not logged in and I try to access an internal course page example, I get a blank page and errors (below). This page should also redirect to the MITx Online login:
home:1 Access to XMLHttpRequest at 'https://rc.mitxonline.mit.edu/oauth2/authorize/?client_id=FV0V9m32EpmIwGaU3MeEruCHWACHgSvolKwpApfi&redirect_uri=https%3A%2F%2Fcourses-qa.mitxonline.mit.edu%2Fauth%2Fcomplete%2Fmitxpro-oauth2%2F%3Fredirect_state%3DfxksXc8prcOY7207AeX5xrPr1UE80b5C&state=fxksXc8prcOY7207AeX5xrPr1UE80b5C&response_type=code&scope=user%3Aread' (redirected from 'https://courses-qa.mitxonline.mit.edu/login_refresh') from origin 'https://mfe-qa.mitxonline.mit.edu' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
rc.mitxonline.mit.edu/oauth2/authorize/?client_id=FV0V9m32EpmIwGaU3MeEruCHWACHgSvolKwpApfi&redirect_uri=https%3A%2F%2Fcourses-qa.mitxonline.mit.edu%2Fauth%2Fcomplete%2Fmitxpro-oauth2%2F%3Fredirect_state%3DfxksXc8prcOY7207AeX5xrPr1UE80b5C&state=fxksXc8prcOY7207AeX5xrPr1UE80b5C&response_type=code&scope=user%3Aread:1 Failed to load resource: net::ERR_FAILED
undefined_license_key:1 Failed to load resource: the server responded with a status of 403 (Forbidden)

After an initial look at the logit feels like the issue is related to the MFE redirection which could be a bad configuration somewhere that's blocking the CORS. But in any case, as it's not directly related to the logout ticket so I think we might do it in a separate ticket.
